File: phases/00-setup-and-tooling/04-apis-and-keys/code/first_api_call.ts
// TypeScript companion for phases/00-setup-and-tooling/04-apis-and-keys/docs/en.md.
// Builds the same Messages-shaped request with a deterministic local fixture.
// Network access requires LIVE=1, ANTHROPIC_API_KEY, and ANTHROPIC_MODEL; MOCK=1 always wins.
// Uses Node's standard library and global fetch, with no SDK dependency.
// Reference: https://docs.anthropic.com/en/api/messages.

import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import process from "node:process";

type MessagesRequest = {
  model: string;
  max_tokens: number;
  messages: { role: "user" | "assistant"; content: string }[];
};

type MessagesResponse = {
  content: { type: string; text: string }[];
  usage: { input_tokens: number; output_tokens: number };
};

// .env loader. Same shape every framework follows; we skip a dep to stay
// portable. KEY=VALUE per line, # comments, optional surrounding quotes.
function loadDotenv(path: string): Record<string, string> {
  let raw: string;
  try {
    raw = readFileSync(path, "utf8");
  } catch {
    return {};
  }
  const out: Record<string, string> = {};
  for (const line of raw.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;
    const eq = trimmed.indexOf("=");
    if (eq <= 0) continue;
    const key = trimmed.slice(0, eq).trim();
    let value = trimmed.slice(eq + 1).trim();
    if (
      (value.startsWith('"') && value.endsWith('"')) ||
      (value.startsWith("'") && value.endsWith("'"))
    ) {
      value = value.slice(1, -1);
    }
    out[key] = value;
  }
  return out;
}

function mergeEnv(): NodeJS.ProcessEnv {
  // process.env wins so users can override the file without editing it.
  const fromFile = loadDotenv(resolve(process.cwd(), ".env"));
  return { ...fromFile, ...process.env };
}

// Fixture matches the real /v1/messages response shape, so the surrounding
// code is identical whether MOCK=1 or not.
const MOCK_RESPONSE: MessagesResponse = {
  content: [
    {
      type: "text",
      text: "A neural network is a stack of differentiable functions that learns patterns by adjusting weights against a loss signal.",
    },
  ],
  usage: { input_tokens: 12, output_tokens: 28 },
};

async function callMessages(
  apiKey: string,
  request: MessagesRequest,
  liveEnabled: boolean,
): Promise<MessagesResponse> {
  if (!liveEnabled || !apiKey || request.model === "lesson-fixture") {
    return MOCK_RESPONSE;
  }

  const resp = await fetch("https://api.anthropic.com/v1/messages", {
    method: "POST",
    headers: {
      "content-type": "application/json",
      "x-api-key": apiKey,
      "anthropic-version": "2023-06-01",
    },
    body: JSON.stringify(request),
  });

  if (!resp.ok) {
    const body = await resp.text();
    throw new Error(`anthropic ${resp.status}: ${body.slice(0, 200)}`);
  }
  return (await resp.json()) as MessagesResponse;
}

async function main(): Promise<number> {
  const env = mergeEnv();
  const apiKey = env.ANTHROPIC_API_KEY ?? "";
  const liveEnabled = env.LIVE === "1";
  const liveModel = env.ANTHROPIC_MODEL?.trim() ?? "";
  const usingMock =
    env.MOCK === "1" || !liveEnabled || !apiKey || !liveModel || liveModel === "lesson-fixture";

  process.stdout.write("=== API Calls ===\n\n");
  process.stdout.write(
    usingMock
      ? "Mode: MOCK (no network). Set LIVE=1, ANTHROPIC_API_KEY, and ANTHROPIC_MODEL for a live call.\n\n"
      : "Mode: LIVE.\n\n",
  );

  const request: MessagesRequest = {
    model: usingMock ? "lesson-fixture" : liveModel,
    max_tokens: 256,
    messages: [{ role: "user", content: "What is a neural network in one sentence?" }],
  };

  try {
    const response = await callMessages(apiKey, request, !usingMock);
    const text = response.content[0]?.text ?? "";
    process.stdout.write(`response: ${text}\n`);
    process.stdout.write(
      `tokens: ${response.usage.input_tokens} in, ${response.usage.output_tokens} out\n`,
    );
    return 0;
  } catch (err) {
    process.stderr.write(`request failed: ${(err as Error).message}\n`);
    return 1;
  }
}

main().then((code) => process.exit(code));

File: phases/00-setup-and-tooling/04-apis-and-keys/code/first_api_call.py
# Lesson implementation for phases/00-setup-and-tooling/04-apis-and-keys/docs/en.md.
# Builds and validates a Messages-shaped request with Python's standard library.
# The default fixture is deterministic and never opens a network connection.
# Raw HTTP is available only with ANTHROPIC_LIVE=1 and ANTHROPIC_API_KEY set.
# Reference: https://docs.anthropic.com/en/api/messages; run with python3 main.py.

from __future__ import annotations

import json
import os
import urllib.request
from collections.abc import Callable, Mapping


API_URL = "https://api.anthropic.com/v1/messages"
API_VERSION = "2023-06-01"
DEFAULT_MODEL = "lesson-fixture"
DEFAULT_PROMPT = "What is a neural network in one sentence?"
DEFAULT_MAX_TOKENS = 64

MOCK_RESPONSE: dict[str, object] = {
    "content": [
        {
            "type": "text",
            "text": "A neural network learns patterns by adjusting weights against a loss signal.",
        }
    ],
    "usage": {"input_tokens": 12, "output_tokens": 17},
}


def build_request(
    prompt: str = DEFAULT_PROMPT,
    *,
    model: str = DEFAULT_MODEL,
    max_tokens: int = DEFAULT_MAX_TOKENS,
) -> dict[str, object]:
    """Return a small, provider-shaped request without sending it."""

    if not prompt.strip():
        raise ValueError("prompt must not be empty")
    if not model.strip():
        raise ValueError("model must not be empty")
    if max_tokens <= 0:
        raise ValueError("max_tokens must be positive")
    return {
        "model": model,
        "max_tokens": max_tokens,
        "messages": [{"role": "user", "content": prompt}],
    }


def build_headers(api_key: str, *, api_version: str = API_VERSION) -> dict[str, str]:
    """Build live-request headers while refusing to handle an absent secret."""

    if not api_key:
        raise ValueError("api_key is required for a live request")
    return {
        "Content-Type": "application/json",
        "x-api-key": api_key,
        "anthropic-version": api_version,
    }


def mock_response(request: Mapping[str, object]) -> dict[str, object]:
    """Return a copy of the deterministic response fixture."""

    del request
    return json.loads(json.dumps(MOCK_RESPONSE))


def call_raw_http(
    api_key: str,
    request: Mapping[str, object],
    *,
    opener: Callable[..., object] = urllib.request.urlopen,
) -> dict[str, object]:
    """Send one explicitly requested HTTP call; tests can inject an opener."""

    model = request.get("model")
    if not isinstance(model, str) or not model.strip() or model == DEFAULT_MODEL:
        raise ValueError("a real ANTHROPIC_MODEL is required for a live request")
    body = json.dumps(request).encode("utf-8")
    http_request = urllib.request.Request(
        API_URL,
        data=body,
        headers=build_headers(api_key),
        method="POST",
    )
    with opener(http_request, timeout=10) as response:
        return json.loads(response.read().decode("utf-8"))


def print_response(response: Mapping[str, object]) -> None:
    content = response["content"]
    usage = response["usage"]
    text = content[0]["text"]  # type: ignore[index]
    input_tokens = usage["input_tokens"]  # type: ignore[index]
    output_tokens = usage["output_tokens"]  # type: ignore[index]
    print(f"response: {text}")
    print(f"tokens: {input_tokens} in, {output_tokens} out")


def main() -> int:
    print("=== API Request Lab ===")

    live_requested = os.environ.get("ANTHROPIC_LIVE") == "1"
    api_key = os.environ.get("ANTHROPIC_API_KEY", "")
    live_model = os.environ.get("ANTHROPIC_MODEL", "").strip()
    if not live_requested:
        request = build_request()
        print(f"request: {json.dumps(request, sort_keys=True)}")
        print("mode: MOCK (no network; set ANTHROPIC_LIVE=1 to opt in)")
        print_response(mock_response(request))
        return 0
    missing = []
    if not api_key:
        missing.append("ANTHROPIC_API_KEY")
    if not live_model:
        missing.append("ANTHROPIC_MODEL")
    if missing or live_model == DEFAULT_MODEL:
        request = build_request()
        print(f"request: {json.dumps(request, sort_keys=True)}")
        reason = (
            f"missing {', '.join(missing)}"
            if missing
            else "ANTHROPIC_MODEL must name a provider model, not lesson-fixture"
        )
        print(f"mode: MOCK (live mode requested, but {reason})")
        print_response(mock_response(request))
        return 0

    request = build_request(model=live_model)
    print(f"request: {json.dumps(request, sort_keys=True)}")
    print("mode: LIVE (explicit opt-in)")
    try:
        print_response(call_raw_http(api_key, request))
    except Exception as error:  # network/provider errors are reported, not hidden
        print(f"live request failed: {error}")
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
